storage/engine/rocksdb/db.cc

```diff
@@ -1089,66 +1151,145 @@ uint64_t DBApproximateSize(DBEngine* db, DBSlice start, DBSlice end) {
+  return ToDBStatus(rep->Put(options, ToSlice(key), ToSlice(value)));
+}
+
+DBStatus DBBatch::Put(DBSlice key, DBSlice value) {
+  ++updates;
+  batch.Put(ToSlice(key), ToSlice(value));
+  return kSuccess;
+}
+
+DBStatus DBSnapshot::Put(DBSlice key, DBSlice value) {
+  return FmtStatus("unsupported");
+}
+
 DBStatus DBPut(DBEngine* db, DBSlice key, DBSlice value) {
+  return db->Put(key, value);
+}
+
+DBStatus DBEngine::Merge(DBSlice key, DBSlice value) {
   rocksdb::WriteOptions options;
-  return ToDBStatus(db->rep->Put(options, ToSlice(key), ToSlice(value)));
+  return ToDBStatus(rep->Merge(options, ToSlice(key), ToSlice(value)));
+}
+
+DBStatus DBBatch::Merge(DBSlice key, DBSlice value) {
+  ++updates;
+  batch.Merge(ToSlice(key), ToSlice(value));
+  return kSuccess;
+}
+
+DBStatus DBSnapshot::Merge(DBSlice key, DBSlice value) {
+  return FmtStatus("unsupported");
 }
 
 DBStatus DBMerge(DBEngine* db, DBSlice key, DBSlice value) {
-  rocksdb::WriteOptions options;
-  return ToDBStatus(db->rep->Merge(options, ToSlice(key), ToSlice(value)));
+  return db->Merge(key, value);
 }
 
-DBStatus DBGet(DBEngine* db, DBSnapshot* snap, DBSlice key, DBString* value) {
-  std::string tmp;
-  rocksdb::Status s = db->rep->Get(MakeReadOptions(snap), ToSlice(key), &tmp);
-  if (!s.ok()) {
-    if (s.IsNotFound()) {
-      // This mirrors the logic in rocksdb_get(). It doesn't seem like
-      // a good idea, but some code in engine_test.go depends on it.
-      value->data = NULL;
-      value->len = 0;
-      return kSuccess;
-    }
-    return ToDBStatus(s);
+DBStatus DBEngine::Get(DBSlice key, DBString* value) {
+  EngineGetter base(this, NULL, key);
+  return base.Get(value);
+}
+
+DBStatus DBBatch::Get(DBSlice key, DBString* value) {
+  if (updates == 0) {
+    return DBEngine::Get(key, value);
   }
-  *value = ToDBString(tmp);
-  return kSuccess;
+  std::unique_ptr<rocksdb::WBWIIterator> iter(batch.NewIterator());
+  rocksdb::Slice rkey(ToSlice(key));
+  iter->Seek(rkey);
+  EngineGetter base(this, NULL, key);
+  return ProcessDeltaKey(&base, iter.get(), rkey, value);
+}
+
+DBStatus DBSnapshot::Get(DBSlice key, DBString* value) {
+  EngineGetter base(this, this, key);
+  return base.Get(value);
+}
+
+DBStatus DBGet(DBEngine* db, DBSlice key, DBString* value) {
+  return db->Get(key, value);
 }
 
-DBStatus DBDelete(DBEngine* db, DBSlice key) {
+DBStatus DBEngine::Delete(DBSlice key) {
   rocksdb::WriteOptions options;
-  return ToDBStatus(db->rep->Delete(options, ToSlice(key)));
+  return ToDBStatus(rep->Delete(options, ToSlice(key)));
+}
+
+DBStatus DBBatch::Delete(DBSlice key) {
+  ++updates;
+  batch.Delete(ToSlice(key));
+  return kSuccess;
+}
+
+DBStatus DBSnapshot::Delete(DBSlice key) {
+  return FmtStatus("unsupported");
+}
+
+DBStatus DBDelete(DBEngine *db, DBSlice key) {
+  return db->Delete(key);
 }
 
-DBStatus DBWrite(DBEngine* db, DBBatch *batch) {
-  if (batch->updates == 0) {
+DBStatus DBEngine::WriteBatch() {
```

Comment:
Yes, it is unsupported by anything but batch. What you're seeing is a limitation of how we're exposing C++ via a C api. Write now, `DBBatch` is not exposed to the Go code. I could expose it, but doing so would require casting to `DBEngine*` for the various `DB*` functions. I think the current code is cleaner.